js/src/logger.ts

```diff
@@ -367,21 +515,54 @@ export async function init(
     apiUrl,
   });
 
-  const ret = await _initExperiment(project, {
+  return await _initExperiment(project, {
     experimentName: experiment,
     description,
     dataset,
     update,
     baseExperiment,
     isPublic,
   });
-  _state.current_experiment = ret;
-  return ret;
 }
 
+/**
+ * Wrapper over `braintrust.init`, which passes the initialized `Experiment` it to the given callback and closes it afterwards. See `braintrust.init` for full details.
+ *
+ * @param options.setCurrent If true (default), set the currently-active experiment to the newly-created one. Equivalent to calling `braintrust.withCurrent(experiment, callback)`.
+ */
+export async function initWithCallback<R>(
+  project: string,
+  options: Readonly<InitOptions & SetCurrentArg> = {},
+  callback: (experiment: Experiment) => R
+): Promise<R> {
```

Comment:
why not just make the callback an optional 3rd argument to `init`, and do this if it's passed in?

Reply:
I was having trouble figuring out how to do that, because it changes the return type of the function from `Experiment` to `Promise<R>`. And I wasn't having much luck with function overloading in typescript.

Reply:
Ah ok let’s leave as is